spark/dl/src/main/scala/com/intel/analytics/bigdl/utils/caffe/LayerConverter.scala

```diff
@@ -0,0 +1,214 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.intel.analytics.bigdl.utils.caffe
+
+import scala.collection.JavaConverters._
+import caffe.Caffe
+import caffe.Caffe.EltwiseParameter.EltwiseOp
+import caffe.Caffe.{BlobProto, PoolingParameter, _}
+import com.google.protobuf.GeneratedMessage
+import com.intel.analytics.bigdl.nn.Graph._
+import com.intel.analytics.bigdl.nn._
+import com.intel.analytics.bigdl.nn.abstractnn.{AbstractModule, Activity}
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.Table
+
+import scala.collection.mutable.ArrayBuffer
+import scala.reflect.ClassTag
+
+
+class LayerConverter[T: ClassTag](implicit ev: TensorNumeric[T]) extends Converter[T]{
+
+  override protected def fromCaffeConvolution(layer : GeneratedMessage) : Seq[ModuleNode[T]] = {
+    val name = getLayerName(layer)
+    val param = getConvolutionParam(layer).get
+    val group = if (param.getGroup == 0)  1 else param.getGroup
+    val  weightBlob = getBlob(layer, 0).get
+    val nInputPlane = if (weightBlob.hasShape) weightBlob.getShape.getDim(1)
+    else weightBlob.getChannels * group
+    val nOutPlane = if (weightBlob.hasShape) weightBlob.getShape.getDim(0)
+    else weightBlob.getNum
+    var kw = param.getKernelW
+    var kh = param.getKernelH
+    var dw = param.getStrideW
+    var dh = param.getStrideH
+    if (kw ==0 || kh == 0) {
+      kw = param.getKernelSize(0)
+      kh = kw
+    }
+    if (dw == 0 || dh == 0) {
+      if (param.getStrideList.size() != 0) {
+        dw = param.getStride(0)
+        dh = dw
+      } else {
+        // use default values if not found
+        dw = 1
+        dh = 1
+      }
+    }
+    var pw = param.getPadW
+    var ph = param.getPadH
+    if (pw == 0 || ph == 0) {
+      if (param.getPadList.size() != 0) {
+        pw = param.getPad(0)
+        ph = pw
+      }
+    }
+    Seq(SpatialConvolution[T](nInputPlane.toInt, nOutPlane.toInt,
+      kw, kh, dw, dh, pw, ph, group).setName(getLayerName(layer)).apply())
+  }
+
+  override protected def fromCaffeInnerProduct(layer : GeneratedMessage) : Seq[ModuleNode[T]] = {
+    val param = getInnerProductParam(layer).get
+    val withBias = param.getBiasTerm
+    val layerName = getLayerName(layer)
+    val weightBlob = getBlob(layer.asInstanceOf[LayerParameter], 0).get
+    var nInputPlane = 0
+    if (weightBlob.hasShape) {
+      nInputPlane = weightBlob.getShape.getDim(1).toInt
+    }
+    else {
+      nInputPlane = weightBlob.getWidth
+    }
+    val nOutputPlane = param.getNumOutput
+    val linear = Linear[T](nInputPlane, nOutputPlane, withBias = withBias).setName(layerName)
+    val node = linear.apply()
+    if(nInputPlane != nOutputPlane) {
+      // Construct a view layer in between
+      val view = View[T](nInputPlane).apply()
```

Comment:
this is method defined in  AbstractModule to create a module node instead of mudule itself
 def apply(nodes : ModuleNode[T]*): ModuleNode[T] = {
****
}